### PR TITLE
The create Artifact request should have a non-empty version.

### LINKED ISSUE #29: The request should fail when creating an Artifact without a version
The artifact create request must have a non-null version value

### PR DESCRIPTION
**Pull request by maheshkelkar**
Added a cucumber test and check in Artifact class.
Removed the non-null restriction on version from DB representtion.

Fixes #29

### CODE REVIEW

File: features/api/artifact/creating.feature [hunk 1]
```diff
@@ -31,17 +31,18 @@ Feature: Artifact CREATE APIs
       }
     """
 
-  @error @wip
-  Scenario: Creating an artifact without a version
+  @error 
+  Scenario: The reuqest should fail when creating an Artifact without a version
```
Comment (rtyler) [inline features/api/artifact/creating.feature]: `request` typo

File: src/main/resources/db/migration/V2__artifacts_versions.sql [hunk 2]
```diff
@@ -8,7 +8,7 @@ ALTER TABLE artifacts
         /*
          * 255 characters of version information should be enough for anybody!
          */
-        version VARCHAR(255) NOT NULL,
+        version VARCHAR(255),
```
Comment (rtyler) [inline src/main/resources/db/migration/V2__artifacts_versions.sql]: Why is this change here? there should be no changes to the database layer for this.

Comment (maheshkelkar) [inline src/main/resources/db/migration/V2__artifacts_versions.sql]: Trying to keep "version" & "sourceUrl" definitions consistent. I had added following comment in the commit description "Removed the non-null restriction on version from DB representtion.".

File: features/api/artifact/creating.feature (hunk 1 above)
Comment (maheshkelkar) [inline features/api/artifact/creating.feature]: will fix and push again.

File: src/main/resources/db/migration/V2__artifacts_versions.sql (hunk 2 above)
Comment (rtyler) [inline src/main/resources/db/migration/V2__artifacts_versions.sql]: This is a database constraint that needs to stay in place, a row with a NULL version is invalid

Comment (maheshkelkar) [inline src/main/resources/db/migration/V2__artifacts_versions.sql]: Why not same constraint on the sourceUrl? Yesterday we deemed it
unnecessary, so I was wondering why not make them consistent. I will back
it out.

Comment (rtyler) [inline src/main/resources/db/migration/V2__artifacts_versions.sql]: I could be misremembering, I think I said we shouldn't be doing request validation with database constraints . We should probably add a subsequent issue to update the database constraint for `sourceUrl`

**Comment by maheshkelkar** [comment]
Apparently code review fixes for #29, somehow are coupled with of fix for #30.

The fix is ready for merge. Can somebody please take care of it?

File: src/cucumber/groovy/step_definitions/HttpSteps.groovy [hunk 3]
```diff
@@ -25,11 +25,37 @@ When(~/^I PUT to "(.*?)" with:$/) { String path, String requestBody ->
...
+When(~/^I PUT to "(.*?)" with a (.*?) over ([1-9][0-9]*) characters$/) { String path, String var, int varSize ->
+
+    // Create a randomString of size varSize+100 
+    String alphabet = (('a'..'z') + ('A'..'Z')+('0'..'9')).join()
+    String randomString = ""
+    for ( i in 1..varSize+100 ) {
+        randomString += alphabet[ new Random().nextInt(alphabet.length()-1)]
+    }
+
+    print "MVK : $var\n"
```
Comment (rtyler) [inline src/cucumber/groovy/step_definitions/HttpSteps.groovy]: This print dshould not be here.

File: src/cucumber/groovy/step_definitions/HttpSteps.groovy [hunk 4]
```diff
@@ -25,11 +25,37 @@ When(~/^I PUT to "(.*?)" with:$/) { String path, String requestBody ->
     response = putJsonToPath(path, requestBody)
 }
 
+When(~/^I PUT to "(.*?)" with a (.*?) over ([1-9][0-9]*) characters$/) { String path, String var, int varSize ->
```
Comment (rtyler) [inline src/cucumber/groovy/step_definitions/HttpSteps.groovy]: This is a *very* specialized step definition just for the artifact creation API which makes it a bit smelly to me.

What we're actually doing in this step is what shohuld be done primarily in the Given step, setting up the test state. I suggest refactoring this into:

```cucumber
Given I have an artifact
 And the group is over 8192 characters
When I PUT ...
Then the response should be 422
```

File: src/cucumber/groovy/step_definitions/HttpSteps.groovy [hunk 5]
```diff
@@ -25,11 +25,37 @@ When(~/^I PUT to "(.*?)" with:$/) { String path, String requestBody ->
...
+
+    // Create a randomString of size varSize+100 
+    String alphabet = (('a'..'z') + ('A'..'Z')+('0'..'9')).join()
```
Comment (rtyler) [inline src/cucumber/groovy/step_definitions/HttpSteps.groovy]: This code is unnecessarily too fancy, just create a big ass string of 'a' characters or something :)

File: src/main/groovy/deploydb/models/Artifact.groovy [hunk 6]
```diff
@@ -16,19 +16,26 @@ import org.hibernate.validator.constraints.NotEmpty
 @Table(name='artifacts')
 class Artifact extends AbstractModel {
 
+    @NotEmpty
+    @Size(min=1, max=8192)
```
Comment (rtyler) [inline src/main/groovy/deploydb/models/Artifact.groovy]: Does `@Size(min=1)` make the `@NotEmpty` annotation redundant now?

Comment (maheshkelkar) [inline src/main/groovy/deploydb/models/Artifact.groovy]: > This print should not be here.
[mvk] Done. Removed

> This is a *very* specialized step definition just for the artifact
creation API which makes it a bit smelly to me.

> What we're actually doing in this step is what should be done primarily
in the Given step, setting up the test state. I suggest refactoring this
into:

> Given I have an artifact
>  And the group is over 8192 characters
> When I PUT ...
> Then the response should be 422

[mvk] The goal is to create one of the attributes with string of 8192+
chars. "Given I have an artifact" creates an artifact, we dont want to
do that. We want to to tweak the requestBody string for an attribute.

[mvk] I assume the parser, parses one step at a time, hence I tried to
compress entire statement in one.

> This code is unnecessarily too fancy, just create a big ass string of 'a' characters or something :)

[mvk] ok. done

Does @Size(min=1) make the @NotEmpty annotation redundant now?

[mvk] no. I tried. Apparently @Size only kicks in if object exists.

**Comment by rtyler** [comment]
@maheshkelkar responding to code comments inline in email doesn't work out very well, better to use the web interface for it :/

**Comment by rtyler** [comment]
I should also mention, that this pull request has an incorrect title which makes it quite confusing

File: src/cucumber/groovy/step_definitions/HttpSteps.groovy (hunk 4 above)
Comment (maheshkelkar) [inline src/cucumber/groovy/step_definitions/HttpSteps.groovy]: We want to create one of the attributes (group, name, version, or sourceUrl) in the requestBody with a string of 8192+ chars. 

"Given I have an artifact" creates an artifact. So I assum we dont want to do that. We want to to tweak the requestBody string for an attribute. 

I assume the parser, parses one step at a time, hence I tried to compress entire statement in one. 

How about the following?
E.g. 
When I PUT to "/api/v1/artifacts" 
  And the group is over 8192 characters

**Comment by maheshkelkar** [comment]
@rtyler Pull request for #29 and #33 somehow got merged under this. Perhaps my comment in #33 commit was "Fixes #33" and I think it should have been "Fix #33". Is there a way to fix this?

File: src/cucumber/groovy/step_definitions/HttpSteps.groovy (hunk 4 above)
Comment (rtyler) [inline src/cucumber/groovy/step_definitions/HttpSteps.groovy]: @maheshkelkar you're right, "Given I have an artifact" is not what we want since it's putting things in the DB. Let's add a new Given step like: "Given I have an artifact request" e.g.:

```cucumber
Given I have an artifact request
And the group is over 8192 characters
When I PUT ..
```

I think it's pretty important to keep the "setup" in the Given stanza

Comment (maheshkelkar) [inline src/cucumber/groovy/step_definitions/HttpSteps.groovy]: *For cucumber:*

```
  @error
  Scenario: Creating an artifact with too large of a group name 2

    Given I have an artifact request
    And the group is over 8192 characters    
    When I PUT to "/api/v1/artifacts"
    Then the response should be 422
```
*Steps*:
```
Given(~/^I have an artifact request$/) { ->
}

And(~/^the (.*?) is over ([1-9][0-9]*) characters$/) { String var, int varSize ->

    // Create a randomString of size varSize+1
    String randomString = "test-".padRight(varSize+1, "a")

    group = var == "group" ? randomString : "com.example.cucumber"
    name  = var == "name" ? randomString : "cukes"
    version = var == "version" ? randomString : "1.2.345"
    sourceUrl = var == "sourceUrl" ? randomString : "http://example.com/cucumber.jar"
    
    requestBody = """ {
        "group" : "$group",
        "name" : "$name",
        "version" : "$version",
        "sourceUrl" : "$sourceUrl"
      }
    """    
}    

When(~/^I PUT to "(.*?)"$/) { String path ->
    response = putJsonToPath(path, requestBody)
}
```

File: src/cucumber/groovy/step_definitions/HttpSteps.groovy (hunk 5 above)
Comment (maheshkelkar) [inline src/cucumber/groovy/step_definitions/HttpSteps.groovy]: done
```
    String randomString = "test-".padRight(varSize+1, "a")
```

**Comment by rtyler** [comment]
I'm going to merge this and @maheshkelkar is going to promise to submit better pull requests with far less git branch confusion in the future
